test/Spec.hs

```diff
@@ -56,7 +58,82 @@ validConcurrencySpec = do
+
+        let cardanoFeatureStart :: forall m. (MonadIO m) => m ()
+            cardanoFeatureStart = generateSingleCommand
+
+        let cardanoFeatureShutdown :: forall m. (MonadIO m) => m ()
+            cardanoFeatureShutdown  = generateSingleCommand
+
+        pure $ CardanoFeature
+            cardanoFeatureName
+            cardanoFeatureStart
+            cardanoFeatureShutdown
+      where
+        -- | Generate random ascii string
+        genSafeText :: Gen Text
+        genSafeText = strConv Lenient <$> listOf1 arbitraryASCIIChar
+
+-- | The function that generates a single command in @IO@.
+generateSingleCommand :: (MonadIO m) => m ()
+generateSingleCommand = do
+    generatedCommand <- liftIO $ generate concurrencyDSLGenerator
+    interepretConcurrencyDSL generatedCommand
+
+-- | The function that creates a list of commands in @IO@.
+generateListOfCommands :: (MonadIO m) => m ()
+generateListOfCommands = do
+    generatedCommands <- liftIO $ generate concurrencyDSLSequenceGenerator
+    void $ traverse interepretConcurrencyDSL generatedCommands
+
+-- | The interpretation of the DSL result. It's been simplified so we don't need to
+-- roam into the type side of the story.
+interepretConcurrencyDSL :: forall m. (MonadIO m) => CommandDSL ->  m ()
+interepretConcurrencyDSL (SequentialCmd delay)  = liftIO . threadDelay $ delay
+interepretConcurrencyDSL (ParallelCmd delay)    = liftIO . void . forkIO . threadDelay $ delay
+interepretConcurrencyDSL (AsyncCmd delay)       = liftIO . void . async . threadDelay $ delay
+
+-- | The delay in microseconds.
+type MicrosecondsDelay = Int
+
+-- | The DSL used to simulate commands. A very general abstraction is that
+-- every command takes some time. For that, we introduce @delay@, we are really not
+-- interested in the command itself or what it does.
+-- We are modeling it's behaviour and are interested in the timings _only_.
+-- We have commands that are sequential, that are parallel and those that are async.
+-- We can simulate all three of these with a simple delay on them.
+-- Again, the delay signifies an actual execution of the command which takes time.
+data CommandDSL
+    = SequentialCmd MicrosecondsDelay
+    | ParallelCmd MicrosecondsDelay
+    | AsyncCmd MicrosecondsDelay
+    deriving (Eq, Show)
+
+-- | The generator of a list of commands, which we can use to
+-- replace a list of commands on.
+concurrencyDSLSequenceGenerator :: Gen [CommandDSL]
+concurrencyDSLSequenceGenerator = listOf1 concurrencyDSLGenerator
+
+-- | The DSL generator. It uses specific frequencies so it a bit biased,
+-- but I think it's showing a more realistic generation like this.
+concurrencyDSLGenerator :: Gen CommandDSL
+concurrencyDSLGenerator = frequency
+    [ (5, SequentialCmd <$> generateMillisecondDelay)
+    , (3, ParallelCmd   <$> generateMillisecondDelay)
+    , (2, AsyncCmd      <$> generateMillisecondDelay)
+    ]
+ where
+    -- | A delay from one millisecond to one whole day.
```

Comment:
one to ten millisecond?

Reply:
On this note, what would be a good interval for testing various features? It seems like it should be grouped into separate sets for testing (1-10 ms, 1000-10000, 10000-100000 for example). A whole day might be problematic for testing purposes as well, as running the tests may take too long? @ksaric what are your thoughts?

Reply:
@robcohen Since this test is actually being checked by DejaFu (or the intention behind is that we test it using it), my understanding is that a large range/interval covering our domain actual values would be ideal, but it turns out that if you increase the delay, the tests slow down.
Ideally, I would like to sit down and write something like this myself and tell you "it doesn't matter", but these things are really complicated and take up a ton of time.
To sum up:
- ideally, you should use the actual domain values (if you have up to one day then one millisecond up to one day)
- in this case, using larger delay slows the computation significantly. 

The case where the slowdown appears in cases when there are too many modules is understandable (combinatorial explosion), but in these cases with delay, it's not really clear to me why this is happening and it could be my lack of understanding how the underlying library works.